hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/EnhancerImpl.java

```diff
@@ -397,19 +407,101 @@ else if ( enhancementContext.doExtendedEnhancement( managedCtClass ) ) {
 		}
 	}
 
+	/**
+	 * Utility that determines the access-type of a mapped class based on an explicit annotation
+	 * or guessing it from the placement of its identifier property. Implementation should be
+	 * aligned with {@code InheritanceState#determineDefaultAccessType()}.
+	 *
+	 * @return the {@link AccessType} used by the mapped class
+	 *
+	 * @implNote this does not fully account for embeddables, as they should inherit the access-type
+	 * from the entities they're used in - defaulting to PROPERTY to always run the accessor check
+	 */
+	private static AccessType determineDefaultAccessType(TypeDefinition typeDefinition) {
+		for ( TypeDefinition candidate = typeDefinition; candidate != null && !candidate.represents( Object.class ); candidate = candidate.getSuperClass() ) {
+			final AnnotationList annotations = candidate.asErasure().getDeclaredAnnotations();
+			if ( hasMappingAnnotation( annotations ) ) {
+				final AnnotationDescription.Loadable<Access> access = annotations.ofType( Access.class );
+				if ( access != null ) {
+					return access.load().value();
+				}
+			}
+		}
+		// Guess from identifier.
+		// FIX: Shouldn't this be determined by the first attribute (i.e., field or property) with annotations,
+		// but without an explicit Access annotation, according to JPA 2.0 spec 2.3.1: Default Access Type?
+		for ( TypeDefinition candidate = typeDefinition; candidate != null && !candidate.represents( Object.class ); candidate = candidate.getSuperClass() ) {
+			final AnnotationList annotations = candidate.asErasure().getDeclaredAnnotations();
+			if ( hasMappingAnnotation( annotations ) ) {
+				for ( FieldDescription ctField : candidate.getDeclaredFields() ) {
+					if ( !Modifier.isStatic( ctField.getModifiers() ) ) {
+						final AnnotationList annotationList = ctField.getDeclaredAnnotations();
+						if ( annotationList.isAnnotationPresent( Id.class ) || annotationList.isAnnotationPresent( EmbeddedId.class ) ) {
```

Comment:
Should we check for any annotation based on a jakarta.persistence class to include other annotations besides ID/EmeddedId?  

List of what I think that would include if done so:

> cd jpa-api/api/src/main/java/jakarta/persistence
> grep -ri "@Retention.*RUNTIME"
> metamodel/StaticMetamodel.java
> ExcludeSuperclassListeners.java
> PersistenceUnits.java
> Access.java
> AssociationOverride.java
> AssociationOverrides.java
> AttributeOverride.java
> AttributeOverrides.java
> Basic.java
> Cacheable.java
> ColumnResult.java
> ConstructorResult.java
> Convert.java
> Converter.java
> Converts.java
> DiscriminatorColumn.java
> DiscriminatorValue.java
> ElementCollection.java
> Embeddable.java
> Embedded.java
> EmbeddedId.java
> Entity.java
> EntityListeners.java
> EntityResult.java
> Enumerated.java
> EnumeratedValue.java
> ExcludeDefaultListeners.java
> FieldResult.java
> ForeignKey.java
> GeneratedValue.java
> Id.java
> IdClass.java
> Index.java
> Inheritance.java
> JoinColumn.java
> JoinColumns.java
> Lob.java
> ManyToMany.java
> ManyToOne.java
> MapKey.java
> MapKeyClass.java
> MapKeyEnumerated.java
> MapKeyJoinColumn.java
> MapKeyJoinColumns.java
> MapKeyTemporal.java
> MappedSuperclass.java
> NamedAttributeNode.java
> NamedEntityGraph.java
> NamedEntityGraphs.java
> NamedNativeQueries.java
> NamedNativeQuery.java
> NamedQueries.java
> NamedQuery.java
> NamedStoredProcedureQueries.java
> NamedStoredProcedureQuery.java
> NamedSubgraph.java
> OneToMany.java
> OneToOne.java
> OrderBy.java
> OrderColumn.java
> PersistenceContext.java
> PersistenceContexts.java
> PersistenceProperty.java
> PersistenceUnit.java
> PostLoad.java
> PostPersist.java
> PostRemove.java
> PostUpdate.java
> PrePersist.java
> PreRemove.java
> PreUpdate.java
> PrimaryKeyJoinColumn.java
> PrimaryKeyJoinColumns.java
> QueryHint.java
> SecondaryTables.java
> SequenceGenerator.java
> SequenceGenerators.java
> SqlResultSetMapping.java
> SqlResultSetMappings.java
> StoredProcedureParameter.java
> TableGenerator.java
> TableGenerators.java
> Temporal.java
> Transient.java
> UniqueConstraint.java
> Version.java
> CheckConstraint.java
> CollectionTable.java
> Column.java
> JoinTable.java
> MapKeyColumn.java
> MapsId.java
> SecondaryTable.java
> Table.java

Many of the ^ don't make sense I guess but maybe there could be other annotations from this ^ list that could make sense to also include?  Maybe IdClass shuld be included as well?  Not a big deal now though.  Good change!

Reply:
This is the same we do during Hibernate bootstrap, see https://github.com/hibernate/hibernate-orm/blob/445f95fb62574caa897338321ed04bab8ad095bc/hibernate-core/src/main/java/org/hibernate/boot/model/internal/InheritanceState.java#L263-L294

We do have a comment there with the intention to account for other annotations, though I believe for now we should keep this aligned.